crypto/src/main/kotlin/net/corda/v5/crypto/merkle/MerkleTreeFactory.kt

```diff
@@ -0,0 +1,91 @@
+package net.corda.v5.crypto.merkle
+
+import net.corda.v5.crypto.DigestAlgorithmName
+import net.corda.v5.base.annotations.DoNotImplement
+import net.corda.v5.base.annotations.Suspendable
+
+/**
+ * An implementation of rfc6962 compatible merkle tree
+ */
+const val HASH_DIGEST_PROVIDER_DEFAULT_NAME = "DefaultHashDigestProvider"
+
+/**
+ * A merkle tree digest that one-way derives a set of per-leaf nonces from a 16-byte entropy input. The nonces are
+ * pre-hashed with the leaf data before forming a standard merkle tree.
+ * This allows low entropy inputs to be used in the merkle tree.
+ * Requires [HASH_DIGEST_PROVIDER_ENTROPY_OPTION]
+ */
+const val HASH_DIGEST_PROVIDER_NONCE_NAME = "NonceHashDigestProvider"
+
+/**
+ * When verifying a NONCE digest based MerkleProof the source entropy is not communicated to the verifier.
+ * Instead, only the per-leaf nonces are transmitted.
+ * Use the NONCE_VERIFY digest to validate the MerkleProof against the expected root hash.
+ */
+const val HASH_DIGEST_PROVIDER_NONCE_VERIFY_NAME = "NonceHashDigestProviderVerify"
+
+/**
+ * Using the NONCE digest it is possible to create a MerkleProof that doesn't reveal the leaf data, but
+ * provides supporting evidence of the number of leaves.
+ * These MerkleProofs can be validated against an expected root hash using the NONCE_SIZE_ONLY_VERIFY digest
+ */
+const val HASH_DIGEST_PROVIDER_NONCE_SIZE_ONLY_VERIFY_NAME = "NonceHashDigestProviderSizeOnlyVerify"
```

Comment:
I think we can't do size only proofs via the public api as that used NonceHashDigestProvider's getSizeProof method. Need to think if we can resolve that via a HASH_DIGEST_PROVIDER_NONCE_SIZE_ONLY_NAME digest, or some other mechanism.

Reply:
I have added a simple extension interface for now.